app/services/credit_notes/refunds/adyen_service.rb

```diff
@@ -0,0 +1,154 @@
+
+        adyen_result = create_adyen_refund
+
+        refund = Refund.new(
+          credit_note:,
+          payment:,
+          payment_provider: payment.payment_provider,
+          payment_provider_customer: payment.payment_provider_customer,
+          amount_cents: adyen_result.response.dig('amount', 'value'),
+          amount_currency: adyen_result.response.dig('amount', 'currency'),
+          status: 'pending',
+          provider_refund_id: adyen_result.response['pspReference'],
+        )
+        refund.save!
+
+        update_credit_note_status(refund.status)
+        track_refund_status_changed(refund.status)
+
+        result.refund = refund
+        result
+      end
+
+      def update_status(provider_refund_id:, status:, metadata: {})
+        refund = Refund.find_by(provider_refund_id:)
+        return handle_missing_refund(metadata) unless refund
+
+        result.refund = refund
+        @credit_note = result.credit_note = refund.credit_note
+        return result if refund.credit_note.succeeded?
+
+        refund.update!(status:)
+        update_credit_note_status(status)
+        track_refund_status_changed(status)
+
+        if status.to_sym == :failed
+          deliver_error_webhook(message: 'Payment refund failed', code: nil)
+          result.service_failure!(code: 'refund_failed', message: 'Refund failed to perform')
+        end
+
+        result
+      rescue ArgumentError
+        result.single_validation_failure!(field: :refund_status, error_code: 'value_is_invalid')
+      end
+
+      private
+
+      attr_accessor :credit_note
+
+      delegate :organization, :customer, :invoice, to: :credit_note
+
+      def client
+        @client ||= Adyen::Client.new(
+          api_key: payment.payment_provider.api_key,
+          env: payment.payment_provider.environment,
+          live_url_prefix: payment.payment_provider.live_prefix
+        )
+      end
+
+      def should_process_refund?
+        return false unless credit_note.refunded?
+        return false if credit_note.succeeded?
+
+        payment.present?
+      end
+
+      def payment
+        @payment ||= credit_note.invoice.payments.order(created_at: :desc).first
+      end
+
+      def adyen_api_key
+        organization.adyen_payment_provider.secret_key
+      end
+
+      def create_adyen_refund
+        result = client.checkout.modifications_api.refund_captured_payment(
+          adyen_refund_params,
+          payment.provider_payment_id
+        )
+      rescue Adyen::AdyenError => e
+        deliver_error_webhook(message: e.msg, code: e.code)
+        update_credit_note_status(:failed)
+
+        raise
+      end
+
+      def adyen_refund_params
+        {
+          paymentPspReference: payment.provider_payment_id,
+          merchantAccount: payment.payment_provider.merchant_account,
+          amount: {
+            value: credit_note.refund_amount_cents,
+            currency: credit_note.credit_amount_currency.upcase
+          }
```

Comment:
Style/TrailingCommaInHashLiteral: Put a comma after the last item of a multiline hash.